pyro/nn/module.py

```diff
@@ -826,3 +827,30 @@ def __set__(self, obj, value):
 
 
 PyroModule[torch.nn.RNNBase]._flat_weights = _FlatWeightsDescriptor()
+
+
+# pyro module list
+# using pyro.nn.PyroModule[torch.nn.ModuleList] can cause issues when
+# slice-indexing nested PyroModuleLists, so we define a separate PyroModuleList
+# class that overwrites the __getitem__ method to return a torch.nn.ModuleList
+# to not use self.__class__ in __getitem__, as that would call the
+# PyroModule.__init__ without the parent module context, leading to a loss
+# of the parent module's _pyro_name, and eventually, errors during sampling
+# as parameter names may not be unique anymore
+# The scenario is rare but happend.
+# The fix could not be applied in torch directly, which is why we have to deal
+# with it here, see https://github.com/pytorch/pytorch/issues/121008
+class PyroModuleList(PyroModule, torch.nn.ModuleList):
+    def __init__(self, modules):
+        PyroModule.__init__(self)
+        torch.nn.ModuleList.__init__(self, modules)
```

Comment:
Can we omit the `__init__` method by swapping inheritance order to be closer to `PyroModule[-]`'s order (defined [here](https://github.com/pyro-ppl/pyro/blob/490a7ff332f65700dd1b43802502f506b8390da3/pyro/nn/module.py#L251-L255))?
```diff
- class PyroModuleList(PyroModule, torch.nn.ModuleList):
+ class PyroModuleList(torch.nn.ModuleList, PyroModule):
```
I think that will be safer with `super.__init__()` handling of subclasses. This will also inherit `torch.nn.ModuleList`'s type hints and default args (`modules` defaults to `None`).

Reply:
Sound good, I applied it in 360fa10

Reply:
Applied in 5ffd8f7.